File: engram-web/lib/arweave.ts
/**
 * Arweave upload utility for Engram
 *
 * Env vars:
 *   ARWEAVE_KEY   — JWK wallet JSON (generate with: node -e "require('arweave').init({}).wallets.generate().then(k=>console.log(JSON.stringify(k)))")
 *   ARWEAVE_ENV   — "mainnet" (default) | "devnet"
 *
 * Devnet uses arweave.net with a test JWK — data is temporary, no AR tokens needed.
 * Mainnet requires a funded wallet (~$0.01 per MB at current AR prices).
 */

import Arweave from "arweave";
import type { JWKInterface } from "arweave/node/lib/wallet";
import { createHash } from "crypto";

const IS_DEVNET = process.env.ARWEAVE_ENV === "devnet";

const arweave = Arweave.init(
  IS_DEVNET
    ? { host: "arweave.net", port: 443, protocol: "https" }
    : { host: "arweave.net", port: 443, protocol: "https" }
);

function getKey(): JWKInterface {
  const raw = process.env.ARWEAVE_KEY;
  if (!raw) throw new Error("ARWEAVE_KEY env var not set");
  try {
    return JSON.parse(raw) as JWKInterface;
  } catch {
    throw new Error("ARWEAVE_KEY is not valid JSON");
  }
}

export function contentCid(buffer: Buffer): string {
  return "sha256:" + createHash("sha256").update(buffer).digest("hex");
}

export interface ArweaveUploadResult {
  tx_id: string;
  url: string;
  content_cid: string;
  size: number;
}

export async function uploadToArweave(
  buffer: Buffer,
  contentType: string,
  tags: Record<string, string> = {}
): Promise<ArweaveUploadResult> {
  const key: JWKInterface = getKey();

  const tx = await arweave.createTransaction({ data: buffer }, key);

  // Standard content-type tag so gateways serve it correctly
  tx.addTag("Content-Type", contentType);
  tx.addTag("App-Name", "Engram");
  tx.addTag("App-Version", "1.0");

  // Custom tags — queryable via Arweave GraphQL
  for (const [k, v] of Object.entries(tags)) {
    tx.addTag(k, v.slice(0, 128)); // Arweave tag value limit
  }

  await arweave.transactions.sign(tx, key);

  const res = await arweave.transactions.post(tx);
  if (res.status !== 200 && res.status !== 202) {
    throw new Error(`Arweave upload failed: status ${res.status}`);
  }

  const tx_id = tx.id;
  const url = `https://arweave.net/${tx_id}`;

  return {
    tx_id,
    url,
    content_cid: contentCid(buffer),
    size: buffer.length,
  };
}

/**
 * Generate a fresh Arweave wallet and return the JWK + address.
 * Used once during setup — store the JWK as ARWEAVE_KEY env var.
 */
export async function generateWallet(): Promise<{ key: JWKInterface; address: string }> {
  const key = await arweave.wallets.generate();
  const address = await arweave.wallets.jwkToAddress(key);
  return { key, address };
}

export async function getWalletBalance(): Promise<{ address: string; ar: string }> {
  const key: JWKInterface = getKey();
  const address = await arweave.wallets.jwkToAddress(key);
  const winstons = await arweave.wallets.getBalance(address);
  const ar = arweave.ar.winstonToAr(winstons);
  return { address, ar };
}

File: engram/storage/arweave.py
"""
Engram — Arweave permanent media storage (Python layer).

Mirrors engram-web/lib/arweave.ts so the full subnet (SDK, miner, CLI) can
upload raw media to Arweave, not just the Next.js web frontend.

Env vars:
  ARWEAVE_KEY  — JWK wallet JSON string
                 Generate: node -e "require('arweave').init({}).wallets.generate().then(k=>console.log(JSON.stringify(k)))"
  ARWEAVE_ENV  — "mainnet" (default) | "devnet"

Upload failures are non-fatal — callers catch ArweaveUnavailable and degrade
gracefully (text still stored in Engram; arweave_tx_id absent from metadata).

Requires:
  pip install arweave-python-client
"""

from __future__ import annotations

import hashlib
import json
import os
import tempfile
from dataclasses import dataclass

from loguru import logger


class ArweaveUnavailable(Exception):
    """Raised when Arweave is not configured or the upload fails."""


@dataclass
class ArweaveUploadResult:
    tx_id: str
    url: str
    content_cid: str
    size: int


_GATEWAY = os.getenv("ARWEAVE_ENV", "mainnet")
_GATEWAY_URL = "https://arweave.net"


def is_configured() -> bool:
    """Return True when ARWEAVE_KEY is present in the environment."""
    return bool(os.environ.get("ARWEAVE_KEY"))


def content_cid(data: bytes) -> str:
    """SHA-256 content identifier matching the web layer's contentCid()."""
    return "sha256:" + hashlib.sha256(data).hexdigest()


def upload(
    data: bytes,
    content_type: str,
    tags: dict[str, str] | None = None,
) -> ArweaveUploadResult:
    """
    Upload raw bytes to Arweave and return the transaction result.

    Args:
        data:         Raw bytes to store permanently.
        content_type: MIME type (e.g. "image/jpeg", "application/pdf").
        tags:         Optional Arweave tags — queryable via GraphQL.

    Returns:
        ArweaveUploadResult with tx_id, url, content_cid, size.

    Raises:
        ArweaveUnavailable: ARWEAVE_KEY not set, package missing, or upload failed.
    """
    try:
        import arweave as _ar  # arweave-python-client
    except ImportError:
        raise ArweaveUnavailable(
            "arweave-python-client not installed. Run: pip install arweave-python-client"
        )

    raw_key = os.environ.get("ARWEAVE_KEY")
    if not raw_key:
        raise ArweaveUnavailable("ARWEAVE_KEY env var not set")

    try:
        jwk = json.loads(raw_key)
    except json.JSONDecodeError as exc:
        raise ArweaveUnavailable("ARWEAVE_KEY is not valid JSON") from exc

    # arweave-python-client loads from a file path — write to an ephemeral temp file
    fd, tmp_path = tempfile.mkstemp(suffix=".json")
    try:
        with os.fdopen(fd, "w") as f:
            json.dump(jwk, f)
        wallet = _ar.Wallet(tmp_path)
    finally:
        os.unlink(tmp_path)

    try:
        tx = _ar.Transaction(wallet, data=data)
        tx.add_tag("Content-Type", content_type)
        tx.add_tag("App-Name", "Engram")
        tx.add_tag("App-Version", "1.0")
        for k, v in (tags or {}).items():
            tx.add_tag(k[:128], v[:128])

        tx.sign()
        resp = tx.send()
    except Exception as exc:
        raise ArweaveUnavailable(f"Arweave transaction failed: {exc}") from exc

    if getattr(resp, "status_code", 200) not in (200, 202):
        raise ArweaveUnavailable(
            f"Arweave upload rejected: HTTP {resp.status_code}"
        )

    tx_id = tx.id
    url = f"{_GATEWAY_URL}/{tx_id}"
    cid = content_cid(data)

    logger.info(
        f"Arweave upload OK | tx={tx_id[:16]}… | {len(data):,} bytes | {content_type}"
    )

    return ArweaveUploadResult(tx_id=tx_id, url=url, content_cid=cid, size=len(data))


def try_upload(
    data: bytes,
    content_type: str,
    tags: dict[str, str] | None = None,
) -> ArweaveUploadResult | None:
    """
    Best-effort upload — returns None instead of raising on any failure.

    Use this in ingest paths where Arweave is supplementary:
      result = arweave.try_upload(pdf_bytes, "application/pdf", {"File-Name": name})
      if result:
          meta["arweave_tx_id"] = result.tx_id
    """
    if not is_configured():
        return None
    try:
        return upload(data, content_type, tags)
    except ArweaveUnavailable as exc:
        logger.warning(f"Arweave upload skipped (non-fatal): {exc}")
        return None
    except Exception as exc:
        logger.warning(f"Arweave upload failed (non-fatal): {exc}")
        return None
